### PR TITLE
[plugins] improve handling of disconnect in social auth

### PR DESCRIPTION
**Pull request by dcramer**
- add lock around get_backends()
- improve handling of get_backends() (dont re-import for no reason)
- move disconnect view into sentry and remove magical failure scenario

### CODE REVIEW

**Comment by dcramer** [comment]
I still have yet to identify what could fully trigger a failed disconnect, but this fixes some awful behavior.

**Comment by dcramer** [comment]
also to be clear, my only goal is to eliminate silent failures

File: src/social_auth/backends/__init__.py [hunk 1]
```diff
@@ -722,35 +717,30 @@ def get_backends(force_load=False):
...
-
-            if issubclass(backend, SocialAuthBackend):
-                name = backend.name
-                backends = getattr(module, 'BACKENDS', {})
-                if name in backends and backends[name].enabled():
-                    BACKENDSCACHE[name] = backends[name]
+    global BACKENDSCACHE
+
+    with _import_lock:
+        if not BACKENDSCACHE or force_load:
```
Comment (mattrobenolt) [inline src/social_auth/backends/__init__.py]: Can we move the `BACKENDSCACHE` check outside the lock:

```python
if not BACKENDSCACHE or force_load:
    with _import_lock:
        ...
```

This way we only acquire the lock once while populating the cache.

**Review by mattrobenolt — APPROVED** [review]
:fire:

File: src/social_auth/backends/__init__.py (hunk 1 above)
Comment (dcramer) [inline src/social_auth/backends/__init__.py]: :+1:

**Pull request merged**